src/missing.c

```diff
@@ -148,64 +156,57 @@ void col_equal_na(r_obj* x,
 
 #define COL_EQUAL_NA(CTYPE, CBEGIN, IS_MISSING) do { \
   CTYPE const* v_x = CBEGIN(x);                      \
+  r_ssize new_count = 0;                             \
                                                      \
-  for (r_ssize i = 0; i < p_info->size; ++i) {       \
-    if (p_info->p_row_known[i]) {                    \
-      continue;                                      \
-    }                                                \
-                                                     \
-    if (!IS_MISSING(v_x[i])) {                       \
-      v_out[i] = 0;                                  \
-      p_info->p_row_known[i] = true;                 \
-      --p_info->remaining;                           \
-                                                     \
-      if (p_info->remaining == 0) {                  \
-        break;                                       \
-      }                                              \
-    }                                                \
+  for (r_ssize i = 0; i < count; ++i) {              \
+    const r_ssize loc = v_loc[i];                    \
+    v_loc[new_count] = loc;                          \
+    new_count += IS_MISSING(v_x[loc]);               \
   }                                                  \
```

Comment:
This new loop is the real key. Using `+= IS_MISSING()` rather than branching when `if (IS_MISSING())` is `true` is much faster, especially in the "moderate amount of NAs" case, where the branching isn't very predictable.